woocommerce/google-listings-and-ads
# Use Ads currency in report pages

### Changes proposed in this Pull Request:

- Move `formatAmountWithCode` to the shared `.~/utils` folder. (90ccce3)
	Addresses https://github.com/woocommerce/google-listings-and-ads/pull/920#discussion_r697353953
- Display correct Ads' currency in Dashobard's `SummarySection`. (f929745)
	Use code instead of symbol.
- Share the metric formatting logic across all report components. (789877c)
	Move the `"Unavailable"` check and precision setup to the `format-amout.js` util.
- Display Ads' currency code on report pages, (e11cb97)
	where applicable - for the "spend" values.
	
	Remove symbol from currency setting for `<Chart>` as it will not be able to change accordingly.
	Affected by https://github.com/woocommerce/woocommerce-admin/issues/7694

Implements part of https://github.com/woocommerce/google-listings-and-ads/issues/363.


### Screenshots:
#### Before: 


#### After: 
![image](https://user-images.githubusercontent.com/17435/134344308-a1699df5-c71e-4b3c-b85d-cc1a5645292e.png)


### Detailed test instructions:

1. [Set the store currency](https://gla1.test/wp-admin/admin.php?page=wc-settings) to _X_ (PLN)
2. Setup Ads account
3. Change the store currency to _Y_ (EUR)
4. Go to the Dashboard ([`/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fdashboard`](https://gla1.test/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fdashboard) )
5. You should see the "Daily budget" column using store's formatting settings, but display Ads' currency symbol.
6. Go to report pages programs & products
7. You should see "Spend" sections, table & chart data in _X_ currency, and sales in _Y_.



### Changelog entry

> Use Ads' account currency on report pages.


### Additional notes:

0. As it may take a while to fetch the Ads currency, I use the store's formatting setting (decimal delimiter, precision, etc. without a symbol and code) to format numbers eagerly. I believe it improves the UX by delivering the meaningful content ASAP, then loading the details later. Currently, this glitch is barely visible, as the majority of the delay in rendering anything is consumed by other APIs. @j111q  Are you ok with such an effect: 
	![lazy load currency](https://user-images.githubusercontent.com/17435/134352211-9798ebbe-3665-4b50-a705-c2c9c201c84b.gif)
1. Google API provides only the currency code, and our serverside here returns just the symbol, so we use the local store's currency formatting settings, for decimal delimiters, etc. That could be problematic when the ads account uses a currency with higher precision. For example on the local store we use `USD` and the sales are `$ 2,345.67`, while the total spend of `0.00123456 BTC`~= `$ 51.75`  will be displayed as `₿ 0.00`. I'm not sure if we should consider it an edge case, try to fetch more applicable precision from somewhere?
2. Y-axis of the Chart has hard-coded store's currency with removed symbol due to wc-admin bug: [https://github.com/woocommerce/woocommerce-admin/issues/7694](https://github.com/woocommerce/woocommerce-admin/issues/7694)
3. Preferably we would use `@woocommerce/currency`'s  `formatAmount` instead of local util, but due to dependency extraction hell, we cannot use the latest version delivered in [https://github.com/woocommerce/woocommerce-admin/pull/7575](https://github.com/woocommerce/woocommerce-admin/pull/7575). Once, that change is available, we could/should rewrite this PR with the simplified flow.
4. To make the displayed currency clearly stated and unambiguous I use currency codes instead of symbols everywhere. That may seem verbose and clutter the space. Alternatively, we could consider:
	1. Using the symbol in the chart tooltips, as the current code is stated already above.
	2. Using the symbol, only if the store's and Ads' currencies are equal.
	3. Using the symbol, only if the store's and Ads' currencies have different symbols.

	But that would be a tad more complicated to implement and could produce less consistent UI when a merchant would change their store's currency. @j111q are you fine with currency codes everywhere, or would you prefer any of the i.-iii. above?
